third_party/tsl/third_party/implib_so/make_stub.py

```diff
@@ -0,0 +1,92 @@
+"""
+Given a list of symbols, generates a stub.
+
+This code is derived from implib-gen, from the Implib.so project, which is under
+the following license.
+
+The MIT License (MIT)
+
+Copyright (c) 2017-2023 Yury Gribov
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+"""
+
+import argparse
+import configparser
+import os
+import re
+import string
+
+from bazel_tools.tools.python.runfiles import runfiles
+
+r = runfiles.Create()
+
+def main():
+  parser = argparse.ArgumentParser(description="Generates stubs for CUDA libraries.",
+                                   formatter_class=argparse.RawDescriptionHelpFormatter)
+  parser.add_argument('symbols',
```

Comment:
We should probably consider improving upstream functionality instead. AFAICT, it should be fairly straightforward to allow accepting the symbol list instead or always running readelf on the DSO.

Us carrying our own generator script relies on implib's implementation details. While it's not a big deal for a tiny dependency like implib, ability to generate the wrapper based on symbol list is a generally useful functionality.

At the very least we should send them a pull request with this script as a stand-alone tool and let upstream owners figure out whether they want to integrate it or not.

Reply:
Yes. I'll look into sending them a PR.

It's not completely obvious that they would accept such a PR, so let's proceed with this in parallel?

Reply:
Agreed, upstreaming can be done separately from this patch.